src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -124,13 +125,27 @@ private BoundStatement RewriteEnumeratorForEachStatement(BoundForEachStatement n
             // Reference to e.
             BoundLocal boundEnumeratorVar = MakeBoundLocal(forEachSyntax, enumeratorVar, enumeratorType);
 
+            var receiver = ConvertReceiverForInvocation(forEachSyntax, rewrittenExpression, getEnumeratorInfo.Method, enumeratorInfo.CollectionConversion, enumeratorInfo.CollectionType);
+
+            // If the GetEnumerator call is an extension method, then the first argument was left unconverted by initial binding. We want to replace
+            // the first argument with our converted receiver and pass null as the receiver instead.
+            if (getEnumeratorInfo.Method.IsExtensionMethod)
```

Comment:
Are we fixing the behavior or merely moving existing logic?

Reply:
Mainly moving, but this logic wasn't necessary in the base because we did the binding of the arguments right here, rather than in initial binding.

---
In reply to: [549493533](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493533) [](ancestors = 549493533)